src/services/RoomService.ts

```diff
@@ -0,0 +1,184 @@
+import io from '../utils/SocketContext';
+import * as Constants from '../utils/Constants';
+import * as Types from '../utils/Types';
+import * as uuid from 'uuid';
+import { Logger } from './Logger';
+const logger = new Logger('room-service');
+
+export class RoomService {
+    rooms: Types.RoomMap;
+
+    constructor() {
+        this.rooms = {};
+    }
+
+    /**
+     * Gets room object with room id
+     *
+     * @param roomId - room id
+     */
+    getRoom(roomId: string) {
+        return this.rooms[roomId];
+    }
+
+    /**
+     * Creates room and returns room id
+     *
+     * @param ownerId - user id of room creator
+     * @param invited - invited list of users
+     */
+    createRoom(ownerId: string, invited: Types.ConnectedUserMap) {
+        const roomId = uuid.v1(); // Generate random time-based id
+        const newRoom: Types.Room = {
+            roomId: roomId,
+            owner: ownerId,
+            requestSent: false,
+            invited: invited,
+        };
+
+        newRoom.invited[ownerId].accepted = true;
+        this.rooms[roomId] = newRoom;
+        logger.logInfo(`${ownerId} has created room with id ${roomId}`);
+        return roomId;
+    }
+
+    /**
+     * Connects user to room
+     *
+     * @param userId - user id connecting to room
+     * @param roomId - room id user is connecting to
+     */
+    connectToRoom(userId: string, roomId: string) {
+        const room = this.getRoom(roomId);
+        room.invited[userId].accepted = true;
+        logger.logInfo(`${userId} connected to room ${roomId}`);
+        let invitedUserIds = Object.keys(room.invited);
+        invitedUserIds.forEach(socketId => {
+            if (room.invited[socketId].accepted) {
+                io.to(socketId).emit(Constants.ROOM_STATUS, {
+                    type: Constants.USER_CONNECT,
+                    roomId: roomId,
+                    full: invitedUserIds.every(invitedUserId => {
+                        return room.invited[invitedUserId].accepted;
+                    }),
+                    owner: room.owner,
+                    userId: userId,
+                });
+            }
+        });
+    }
+
+    /**
+     * Deletes room
+     *
+     * @param roomId - room id
+     */
+    deleteRoom(roomId: string) {
+        delete this.rooms[roomId];
```

Comment:
Hm, maybe we can use assignment instead of mutation here too? Like
```
return this.rooms.filter(room => room.roomId != roomId);

or

this.rooms = this.rooms.filter(room => room.roomId != roomId);
```

Thoughts?

Reply:
I am not sure what the run time difference of each operation would be but I would think delete is """"faster""""? But if we want to keep things consistent, I agree that assignment would be better.

Reply:
They're still both o(n) tho right? And the assignment is just changing a pointer so that doesnt change anything. I think assignment will be clearer to know when the state is changing.

Reply:
Shouldn't `delete` be o(1) and filter o(n)?

Reply:
Also, can you even filter through an object without creating a custom iterator?